lib/sidekiq_flow/client.rb

```diff
@@ -186,6 +203,46 @@ def find_first(key_pattern)
         result
       end
 
+      # Optimization to avoid N+1 Redis scans
+
+      def timestamp_namespace
+        "workflow-timestamps"
+      end
+
+      def store_start_timestamp(workflow_id, timestamp)
+        connection_pool.with do |redis|
+          redis.set("#{timestamp_namespace}.#{workflow_id}.start", timestamp)
+        end
+      end
+
+      def build_workflow_key_from_timestamps(workflow_id)
+        start_timestamp, end_timestamp = connection_pool.with do |redis|
+          redis.pipelined do
+            redis.get("#{timestamp_namespace}.#{workflow_id}.start")
+            redis.get("#{timestamp_namespace}.#{workflow_id}.end")
+          end
+        end
+
+        workflow_key = if end_timestamp && start_timestamp
+                         "#{configuration.namespace}.#{workflow_id}_#{start_timestamp}_#{end_timestamp}"
+                       elsif start_timestamp
+                         "#{configuration.namespace}.#{workflow_id}_#{start_timestamp}_0"
+                       end
+        # sanity check find in case workflow key was already deleted
+        if workflow_key && workflow_key_exists?(workflow_key)
+          workflow_key
+        else
+          nil
+        end
+      end
+
+      def workflow_key_exists?(workflow_key)
+        connection_pool.with do |redis|
+          # NOTE: Redis ~>4.2.5 modified exists to be a variadic function
+          #       and returns integers.
+          redis.exists(workflow_key)
```

Comment:
Any reason not to use `exists?` which return a Boolean ?

Reply:
Updated. had to upgrade redis to 4.2.5 to make it work in gemspec.